lib/segment/src/index/sparse_index/sparse_vector_index.rs

```diff
@@ -99,18 +105,34 @@ impl<TInvertedIndex: InvertedIndex> SparseVectorIndex<TInvertedIndex> {
         vector_storage: Arc<AtomicRefCell<VectorStorageEnum>>,
         path: &Path,
         stopped: &AtomicBool,
-    ) -> OperationResult<TInvertedIndex> {
+    ) -> OperationResult<(TInvertedIndex, IndicesTracker)> {
         let borrowed_vector_storage = vector_storage.borrow();
         let borrowed_id_tracker = id_tracker.borrow();
         let deleted_bitslice = borrowed_vector_storage.deleted_vector_bitslice();
-        let mut ram_index = InvertedIndexRam::empty();
-        let mut index_point_count: usize = 0;
+
+        let mut indices_tracker = IndicesTracker::default();
         for id in borrowed_id_tracker.iter_ids_excluding(deleted_bitslice) {
             check_process_stopped(stopped)?;
             // It is possible that the vector is not present in the storage in case of crash.
             // Because:
             // - the `id_tracker` is flushed before the `vector_storage`
             // - the sparse index is built *before* recovering the WAL when loading a segment
+            match borrowed_vector_storage.get_vector_opt(id) {
+                None => {
+                    // the vector was lost in a crash but will be recovered by the WAL
+                    log::debug!("Sparse vector with id {} is not found", id)
+                }
+                Some(vector) => {
+                    let vector: &SparseVector = vector.as_vec_ref().try_into()?;
+                    indices_tracker.register_indices(vector);
```

Comment:
can we do a single pass instead? register and remap in one go?

Reply:
Thanks for this observation! Fixed